apiBase/APIRequestUtils.es6.js

```diff
@@ -145,38 +149,60 @@ const normalizeRequest = (res, req) => {
   return req;
 };
 
-const handleRequestIfFailed = (apiOptions, err, res, req, reject) => {
-  if (!(err || (res && !res.ok))) { return; }
-
-  getEmitter(apiOptions).emit(Events.error, err, req);
-  const errorHandler = apiOptions.defaultErrorHandler || reject;
-  errorHandler(err || 500);
-  return true;
-};
-
-const handle = (apiOptions, resolve, reject, err, res, req, query, method,
+const handle = (apiOptions, resolve, reject, err, res, req, method, path, query,
     parseBody, parseMeta) => {
 
   req = normalizeRequest(res, req);
 
-  if (handleRequestIfFailed(apiOptions, err, res, req, reject)) {
+  if (handleRequestIfFailed(apiOptions, err, res, req, method, path, reject)) {
     return;
   }
 
   getEmitter(apiOptions).emit(Events.response, req, res);
 
-  const apiResponse = tryParseResponse(reject, res, req, method, query, parseBody, parseMeta);
+  const apiResponse = tryParseResponse(reject, res, req, method, path, query, parseBody, parseMeta);
 
   getEmitter(apiOptions).emit(Events.result, apiResponse);
   resolve(apiResponse);
 };
 
-const tryParseResponse = (reject, res, req, method, query, parseBody, parseMeta) => {
+const handleRequestIfFailed = (apiOptions, err, res, req, method, path, reject) => {
+  if (!(err || (res && !res.ok))) { return; }
+
+  // Error handling ported from 1X, this could be revisited in more depth
+  // but it seems like getting ResponseErrors to be thrown is more important
+  // for now.
+  if (err) {
+    getEmitter(apiOptions).emit(Events.error, err, req);
+
+    if (err && err.timeout) {
+      err.status = 504;
+    }
+
+    // This is a good example of error-handling we should reconsider soon.
+    // The api sending back a 302 for a subreddit might be a good indicator
+    // that we should really be issuing a 302 to a url that has the canonical
+    // capitalization of a subreddit for SEO or any number of other things.
+    // 1X's comment:
+    // Sometimes the API decides to send back a 302 instead of a 404; for
+    // example, unfound subs will redirect you to the search page.
+    if (method === 'get' && status === 302) {
```

Comment:
It's a bit messy, I did some grepping into r2 for the specific case the 1X comment mentions (a specific edge case when you do subreddit filtering via the url and the subreddits are invalid). That code is used in MinimalController and looks like it can be triggered on any endpoint that has a subreddit name in it. (Realistically its probably just the `r/<subreddit>/about.json` and `r/<subreddit>`). It looks like its an edge case specific to subreddit filtering that we don't support anyways on mweb at the moment, so I'd be in favor of just removing it all together for now and re-adding it in some other form later if it comes up again.